### PR TITLE
Handler formulier conditions refactor

### PR DESCRIPTION
**Pull request by hghanbari**

### CODE REVIEW

**Review by lencodes — CHANGES_REQUESTED** [review]
The made changes are fine. However, they are not complete.

1. We're not sending the data to the gateway (you've created a TextArea of which we never extract the value when saving the `Handler`);

2. When we do start extracting the data and send it to the gateway, the gateway doesn't save it this is because the gateway expects an array of items (e.g. `['condition', 'condition']`); when we push an array, the gateway does save it.

Note on (2.): I think this is an issue on the backend, it shouldn't be accepting nor returning an array. However, you should double check this with Ruben. Also, if the backend is supposed to save and return arrays, then on the frontend it should not be a TextArea. 

Call me if you have any questions.

**Comment by bbrands02** [comment]
Conditions is now a textfield. It gets prefilled with array/json converted to string. Then when posting the input gets parsed back to json. Also an bugfix for the description field is included otherwise this form would not work. The API calls still do fetching the old way but that is a different issue/ticket.

**Review by lencodes — COMMENTED** [review]
Waiting on backend changes.

**Comment by bbrands02** [comment]
image
Backend change has been made and is ready in the frontend

File: pwa/src/components/handlers/handlerForm.tsx [hunk 1]
```diff
@@ -173,13 +178,18 @@ export const HandlerForm: React.FC<HandlerFormProps> = ({ id, endpointId }) => {
...
+
     let url = `${context.adminUrl}/handlers`;
     let method = "POST";
     if (id) {
       url = `${url}/${id}`;
       method = "PUT";
     }
 
+
+    console.log(body);
```
Comment (lencodes) [inline pwa/src/components/handlers/handlerForm.tsx]: Added some unnecessary whitespace and left a `console.log` in.

File: pwa/src/components/handlers/handlerForm.tsx [hunk 2]
```diff
@@ -337,49 +347,52 @@ export const HandlerForm: React.FC<HandlerFormProps> = ({ id, endpointId }) => {
...
                       </div>
                     </div>
+                    <div className="row">
+                      <div className="col-6">
+                        <TextareaGroup
+                          name={"conditions"}
+                          id={"conditionsInput"}
+                          defaultValue={handler?.conditions}
+                          required={true}
+                        />
```
Comment (lencodes) [inline pwa/src/components/handlers/handlerForm.tsx]: There are a couple of issues here.

1. The gateway expects a valid JSON to be sent within the `conditions` textarea, this should reflect in the frontend.
2. The frontend is showing success alerts while getting a `422` response from the gateway (this should be showing the error alert and message).

**Review by lencodes — CHANGES_REQUESTED** [review]
-

File: pwa/src/components/handlers/handlerForm.tsx [hunk 3]
```diff
@@ -296,49 +315,63 @@ export const HandlerForm: React.FC<HandlerFormProps> = ({ handlerId, endpointId
...
+                            return (
+                              tableNames ? (
+                                <MultiSelect
+                                  id="translationsOut"
+                                  label="Translations Out"
+                                  data={handler?.translationsOut}
+                                  options={tableNames}
+                                />
+                              ) : (
+                                <><Spinner /></>
```
Comment (lencodes) [inline pwa/src/components/handlers/handlerForm.tsx]: This doesn't need to be wrapped in `<></>`. It can just be the `<Spinner />` I think.

File: pwa/src/components/handlers/handlerForm.tsx [hunk 4]
```diff
@@ -296,49 +315,63 @@ export const HandlerForm: React.FC<HandlerFormProps> = ({ handlerId, endpointId
...
+                            return (
+                              tableNames ? (
+                                <MultiSelect
+                                  id="translationsIn"
+                                  label="Translations In"
+                                  data={handler?.translationsIn}
+                                  options={tableNames}
+                                />
+                              ) : (
+                                <><Spinner /></>
```
Comment (lencodes) [inline pwa/src/components/handlers/handlerForm.tsx]: This doesn't need to be wrapped in `<></>`. It can just be the `<Spinner />` I think.

File: pwa/src/components/handlers/handlerForm.tsx [hunk 5]
```diff
@@ -296,49 +315,63 @@ export const HandlerForm: React.FC<HandlerFormProps> = ({ handlerId, endpointId
...
                       </div>
                     </div>
+                    <div className="row">
+                      <div className="col-6">
+                        <TextareaGroup
+                          name={"conditions"}
+                          label={"Conditions (JSON)"}
+                          id={"conditionsInput"}
+                          defaultValue={handler?.conditions}
+                          required={true}
```
Comment (lencodes) [inline pwa/src/components/handlers/handlerForm.tsx]: When passing a boolean, you do not need to explicitly set it when it's true.

So:

`required` => defaults to `required={true}`

So only when it's false you have to actually write it out (like `required={false}`)

File: pwa/src/components/handlers/handlerForm.tsx [hunk 6]
```diff
@@ -296,49 +315,63 @@ export const HandlerForm: React.FC<HandlerFormProps> = ({ handlerId, endpointId
...
+                          infoTooltip={{
+                            content: <p>
+                              This field expects a<br/>
+                              <a
+                                target="_blank" href="https://jsonlint.com/"
+                              >
+                                valid JSON object
+                              </a>
+                            </p>
```
Comment (lencodes) [inline pwa/src/components/handlers/handlerForm.tsx]: The pop-up unfortunately seems to go away when I hover it. Resulting in not being able to click the actual link. I suggest removing it altogether for now.

Note: I feel as if we can remove it because you've built a well-working JSON check and error message.

File: pwa/src/components/handlers/handlerForm.tsx [hunk 7]
```diff
@@ -105,12 +115,13 @@ export const HandlerForm: React.FC<HandlerFormProps> = ({ handlerId, endpointId
     // get the inputs and check if set other set null
     let body: {} = {
```
Comment (lencodes) [inline pwa/src/components/handlers/handlerForm.tsx]: Because body is currently of type `{}` it does not have (e.g.) `.conditions` which is giving a Typescript error on line `143` so body should be of type `any`.

Comment (lencodes) [reply pwa/src/components/handlers/handlerForm.tsx]: Edit: solved.

File: pwa/src/components/handlers/handlerForm.tsx (hunk 6 above)
Comment (lencodes) [reply pwa/src/components/handlers/handlerForm.tsx]: Edit: solved.

**Review by lencodes — CHANGES_REQUESTED** [review]
